File: AFQMC/orthogonalization.c
/*This File Contains Routines to Orthogonalize the Columns*/

#include "telluride_basic_cpmc.h"

/****************************************************************************************************/

void orthogonalize(double *wf_up,double *wf_down,int_st ist){

    /*Orthogonalizes Wave Function After Using Random Numbers to Determine To Prevent Overlaps*/

    double *R_up, *R_down;

    R_up = (double *)malloc(ist.n_sites_sq*sizeof(double));
    R_down = (double *)malloc(ist.n_sites_sq*sizeof(double));

    /*Performs Modified Gram Schmidt to Orthogonalize*/
    modified_gram_schmidt(wf_up,R_up,ist.n_sites,ist.n_up);
    modified_gram_schmidt(wf_down,R_down,ist.n_sites,ist.n_down);

free(R_up);
free(R_down);
return;
}

/****************************************************************************************************/

void modified_gram_schmidt(double *q, double *r, int size1, int size2) {

   /*The Modified Gram-Schmidt Routine Used to Stabilize Matrix Products*/
   int i, j, k;
   double temporary, anorm;
   double *d;

   d=(double *)calloc(size2,sizeof(double));
   dzero(r,size2*size2);

   for (i=1; i<=size2; i++) {

     temporary = 0.0;
     for (j=1; j<=size1; j++) {
       temporary += q[(j-1)*size2+i-1]*q[(j-1)*size2+i-1];
     }
     d[i-1]=sqrt(temporary);
     anorm = 1.0/d[i-1];

     for (j=1; j<=size1; j++) {
       q[(j-1)*size2+i-1] *= anorm;
     }

     for ( j=i+1; j<=size2; j++) {
       temporary = 0.0;
       for (k=1; k<=size1; k++) {
         temporary += q[(k-1)*size2+i-1] * q[(k-1)*size2+j-1];
       }

       for (k=1; k<=size1; k++) {
         q[(k-1)*size2+j-1]-= temporary * q[(k-1)*size2+i-1];
       }
       r[(i-1)*size2+j-1]=temporary*anorm;
      }

   } /*i loop*/

   /*Now Make V Into R Matrix*/
   for (i=0; i<size2; i++) {
    r[i*size2+i]=d[i];
   }

   free(d);

return;
}

/*************************************************************************************/

File: Day4_AFQMC/Solutions/orthogonalization.py
import numpy as np

# Orthogonalization of Wave Function Using Modified Gram Schmidt

#----------------------------------------------------------------------------
def orthogonalize(wf_up,wf_down,size,n_up,n_down):
   # Orthogonalize Wave Function to Prevent Collapse of Columns

   # Perform Modified Gram Schmidt on Each Wf
   wf_up = modified_gram_schmidt(wf_up,size,n_up)
   wf_down = modified_gram_schmidt(wf_down,size,n_down)

#----------------------------------------------------------------------------
def modified_gram_schmidt(q,size1,size2):
   # Orthogonalizes Columns of Matrices
  
   # Initialize Diagonal and R Pieces
   d = np.zeros(size2)
   r = np.zeros( (size2,size2) )
   temporary = 0.0
   anorm = 0.0

   # Perform Modified Gram Schmidt on Each Column - Could Be Faster Numpy Way
   for i in range(1,size2+1):
   
     temporary = 0.0
     for j in range(1,size1+1):
      temporary += q[j-1,i-1] * q[j-1,i-1]
     d[i-1]=np.sqrt(temporary)
     anorm = 1.0/d[i-1]

     for j in range(1,size1+1):
      q[j-1,i-1] *= anorm

     for j in range(i+1,size2+1):
      temporary = 0.0
      for k in range(1,size1+1):
         temporary += q[k-1,i-1] * q[k-1,j-1]
     
      for k in range(1,size1+1):
          q[k-1,j-1] -= (temporary * q[k-1,i-1])
      r[i-1,j-1]=temporary*anorm
  
   for i in range(size2):
     r[i,i] = d[i]

   return q

#----------------------------------------------------------------------------
